abdiff/core/collate_ab_transforms.py

```diff
@@ -0,0 +1,220 @@
+    )
+
+    joined_written_files = write_to_dataset(
+        get_joined_batches_iter(transformed_dataset_path.name),
+        base_dir=collated_dataset_path,
+        schema=JOINED_DATASET_SCHEMA,
+    )
+    logger.info(f"Wrote {len(joined_written_files)} parquet file(s) to collated dataset")
+    return collated_dataset_path
+
+
+def yield_records(transformed_file: str | Path) -> Generator:
+    """Yields data for every TIMDEX record in a transformed file.
+
+    This function uses ijson to yield records from a JSON stream
+    (i.e., the transformed file) one at a time. A generator is returned,
+    yielding a dictionary that contains the following:
+
+    * timdex_record_id: The TIMDEX record ID.
+    * source: The shorthand name of the source as denoted in by Transmogrifier
+      (see https://github.com/MITLibraries/transmogrifier/blob/main/transmogrifier/config.py).
+    * record: The TIMDEX record serialized to a JSON string then encoded to bytes.
+    * version: The version of the transform, parsed from the absolute filepath to a
+      transformed file.
+    * transformed_file_name: The name of the transformed file, excluding file extension.
+    """
+    version, transformed_file_name = parse_parquet_details_from_transformed_file(
+        str(transformed_file)
+    )
+    with open(transformed_file) as file:
+        records = ijson.items(file, "item")
+        for record in records:
+            yield {
+                "timdex_record_id": record["timdex_record_id"],
+                "source": record["source"],
+                "record": json.dumps(record).encode(),
+                "version": version,
+                "transformed_file_name": transformed_file_name,
+            }
+
+
+def get_transformed_batches_iter(
+    run_directory: str, transformed_files: tuple[list[str], ...]
+) -> Generator:
+    """Yield pyarrow.RecordBatch objects of TIMDEX records.
+
+    This function will iterate over the A/B lists in 'transformed_files',
+    calling yield_records() to fetch dictionaries describing
+    TIMDEX records and compiling the dictionaries into a
+    pyarrow.RecordBatch. The size of a batch is set by the
+    'READ_BATCH_SIZE' global variable.
+
+    The function returns a generator, yielding batches of the dictionaries
+    from yield_records(). The returned generator can be passed to
+    abdiff.core.utils.write_to_dataset() to perform batch writes to
+    a Parquet dataset.
+    """
+    for transformed_files_list in transformed_files:
+        for transformed_file in transformed_files_list:
+            record_iter = yield_records(
+                transformed_file=Path(run_directory) / transformed_file
+            )
+            for record_batch in itertools.batched(record_iter, READ_BATCH_SIZE):
+                yield pa.RecordBatch.from_pylist(list(record_batch))
+
+
+def get_joined_batches_iter(dataset_directory: str) -> Generator:
```

Comment:
More specific `Generator` type hint

Reply:
Updated return type hint to `Generator[pa.RecordBatch]`.

Reply:
Jumping in here arbitrarily, but appreciate these more detailed type hints requests.  Don't _always_ think they have lots of value, but when we're interacting with a newish library to us, or one that has unusual inputs/outputs, I think it's quite helpful.